### PR TITLE
[MoMo] Support extracting correlation id in MDC preprocessor

### PR DESCRIPTION
**Pull request by chupadhy**

### CODE REVIEW

File: assets/model_monitoring/components/src/model_data_collector_preprocessor/run.py [hunk 1]
```diff
@@ -58,6 +46,31 @@ def mdc_preprocessor(
...
+    data_window_start: str,
+    data_window_end: str,
+    input_data: str,
+    preprocessed_input_data: str,
+    extract_correlation_id: bool
+):
+    """Extract data based on window size provided and preprocess it into MLTable.
+
+    Args:
+        production_data: The data asset on which the date filter is applied.
```
Comment (yetamsft) [inline assets/model_monitoring/components/src/model_data_collector_preprocessor/run.py:62]: Could you please update this Args list? It doesn't seem to match.

File: assets/model_monitoring/components/model_monitor/model_data_collector_preprocessor/spec.yaml [hunk 2]
```diff
@@ -19,6 +19,9 @@ inputs:
   input_data:
     type: uri_folder
     mode: direct
+  extract_correlation_id:
+    type: string
```
Comment (yetamsft) [inline assets/model_monitoring/components/model_monitor/model_data_collector_preprocessor/spec.yaml:23]: nit: I think `type: boolean` is supported in yaml. Right?

**Review by yetamsft — APPROVED** [review]

File: assets/model_monitoring/components/model_monitor/model_data_collector_preprocessor/spec.yaml (hunk 2 above)
Comment (mastloui-msft) [reply assets/model_monitoring/components/model_monitor/model_data_collector_preprocessor/spec.yaml:23]: Reference:
https://learn.microsoft.com/en-us/azure/machine-learning/reference-yaml-job-pipeline?view=azureml-api-2#job-inputs

File: assets/model_monitoring/components/tests/e2e/test_model_data_collector_preprocessor_e2e.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
+# Copyright (c) Microsoft Corporation.
```
Comment (mastloui-msft) [inline assets/model_monitoring/components/tests/e2e/test_model_data_collector_preprocessor_e2e.py:1]: Thoughts on adding the input MDC preprocessor + data joiner as part of the feature attribution drift signal e2e test?

File: assets/model_monitoring/components/src/model_data_collector_preprocessor/run.py [hunk 4]
```diff
@@ -85,17 +98,35 @@ def mdc_preprocessor(
...
+        for df in iterator:
+            yield pd.concat(
+                extract_data_and_correlation_id(getattr(row, MDC_DATA_COLUMN), getattr(row, MDC_CORRELATION_ID_COLUMN)) for row in df.itertuples() # noqa
+            )
+
+    def extract_data_and_correlation_id(entry, correlationid):
+        result = pd.read_json(entry)
+        result[MDC_CORRELATION_ID_COLUMN] = ""
+        for index, row in result.iterrows():
+            result.loc[index, MDC_CORRELATION_ID_COLUMN] = correlationid + "_" + str(index)
```
Comment (mastloui-msft) [inline assets/model_monitoring/components/src/model_data_collector_preprocessor/run.py:120]: oh I see you ended up implementing this right away!

File: assets/model_monitoring/components/src/model_data_collector_preprocessor/run.py [hunk 5]
```diff
@@ -10,40 +10,28 @@
...
     )
 
-    # TODO Validate window size
-    end_datetime = parser.parse(data_window_end)
-    end_datetime = datetime.strptime(
-        str(end_datetime.strftime(format_data)), format_data
-    )
 
-    # Create mltable, create column with partitionFormat
+def _create_mltable_definition_to_transform_mdc_data(
```
Comment (mastloui-msft) [inline assets/model_monitoring/components/src/model_data_collector_preprocessor/run.py:30]: wondering if a name like `convert_uri_folder_to_mltable` would be more fitting here. I wasn't sure what this was doing from the name and had to read through the code to figure it out!

**Review by mastloui-msft — APPROVED** [review]

File: assets/model_monitoring/components/tests/e2e/test_model_data_collector_preprocessor_e2e.py (hunk 3 above)
Comment (chupadhy) [reply assets/model_monitoring/components/tests/e2e/test_model_data_collector_preprocessor_e2e.py:1]: Oh yes we should absolutely do it!! I was thinking we will do it in GSQ but I totally forgot we have FAD. Let me add an item

File: assets/model_monitoring/components/src/model_data_collector_preprocessor/run.py (hunk 4 above)
Comment (chupadhy) [reply assets/model_monitoring/components/src/model_data_collector_preprocessor/run.py:120]: Yes this has all the MDC changes

File: assets/model_monitoring/components/tests/e2e/test_model_data_collector_preprocessor_e2e.py (hunk 3 above)
Comment (chupadhy) [reply assets/model_monitoring/components/tests/e2e/test_model_data_collector_preprocessor_e2e.py:1]: https://msdata.visualstudio.com/Vienna/_workitems/edit/2612466

File: assets/model_monitoring/components/src/model_data_collector_preprocessor/run.py (hunk 5 above)
Comment (chupadhy) [reply assets/model_monitoring/components/src/model_data_collector_preprocessor/run.py:30]: Good idea - will follow-up in a bit. Thanks

File: assets/model_monitoring/components/src/model_data_collector_preprocessor/run.py (hunk 1 above)
Comment (chupadhy) [reply assets/model_monitoring/components/src/model_data_collector_preprocessor/run.py:62]: Thanks I will follow-up in a bit

File: assets/model_monitoring/components/model_monitor/model_data_collector_preprocessor/spec.yaml (hunk 2 above)
Comment (chupadhy) [reply assets/model_monitoring/components/model_monitor/model_data_collector_preprocessor/spec.yaml:23]: I will follow-up

**Pull request merged**